compiler/sem/pragmas.nim

```diff
@@ -1728,60 +1748,171 @@ proc hasPragma*(n: PNode, pragma: TSpecialWord): bool =
+  of nkPragma:
+    result = n
+  else:
+    unreachable(n.kind)
+
+  let start = result.len
+
+  # step 1: add the implicit pragmas to the list
+  var tmp: seq[PNode]
+  for it in c.optionStack.items:
+    let o = it.otherPragmas
+    if o != nil:
+      for x in o.items:
+        prepareSinglePragma(c, x, tmp, validPragmas, sym)
+
+        # filter the nodes. If they're erroneous, it means that the pragma
+        # doesn't apply to the symbol
+        for y in tmp.items:
+          if not y.isError:
+            # we're mutating the pragma's AST, so a copy is required
+            let pragma = copyTree(y)
+            overwriteLineInfo(pragma, n.info)
+            result.add(pragma)
+
+        tmp.setLen(0)
+
+  if result.len == 0:
+    # there were no applicable pragmas; restore the original
+    result = n
+  else:
+    # apply the merged pragmas
+    if applySymbolPragmaList(c, sym,
+                             result.sons.toOpenArray(start, result.len-1)):
+      # application failed, wrap the error nodes with another error indicating
+      # that the pragmas are implicit
+      for it in result.sons.mitems:
+        if it.kind == nkError:
+          it = c.config.newError(it):
+            PAstDiag(kind: adSemImplicitPragmaError, implicitPragma: sym)
+
+      result = c.config.wrapError(result)
+
+proc pragmaDecl*(c: PContext, sym: PSym, n: PNode, validPragmas: TSpecialWords): PNode =
+  ## Similar to `pragmaDeclNoImplicit <#pragmaDeclNoImplicit>`_, but adds the
+  ## implicit pragmas to the resulting list and applies.
+  addInNimDebugUtils(c.config, "pragmaDecl", sym, n, result)
+  result = pragmaDeclNoImplicit(c, sym, n, validPragmas)
+  # only add the implicit pragmas if there was no error during application of
+  # the explicit ones
+  if result.kind != nkError:
+    result = implicitPragmas(c, sym, result, validPragmas)
+
+# NOTE: types previously also inherited a pushed dynlib
+proc inheritDynlib*(c: PContext, sym: PSym) =
+  ## Apply the dynlib pragma from the top of the option stack to `sym`, if
+  ## applicable. The dynlib pragma can be applied if the symbol is marked as
+  ## imported, but no header nor dynlib are specified.
+  let lib = c.optionStack[^1].dynlib
+  if lib != nil and sfImportc in sym.flags and
+     {lfDynamicLib, lfHeader} * sym.loc.flags == {}:
+    incl(sym.loc.flags, lfDynamicLib)
+    addToLib(lib, sym)
+    if sym.loc.r == "":
+      # XXX: this looks like a unnecessary defensive check. If the symbol is
+      #      marked as imported, it already has an external name set
+      sym.loc.r = sym.name.s
+
+proc containsError(n: PNode): bool =
+  for it in n.items:
+    if it.kind == nkError:
+      return true
+
+  result = false
+
+proc pragmaExpr*(c: PContext, n: PNode): PNode =
+  ## Pre-processes the pragma list `n` for the purpose of further processing
+  ## by sem in the context of applying pragmas to an expression.
+  assert n.kind == nkPragma
+  result = copyNode(n)
+  for i in 0..<n.len:
```

Comment:
Just curious why favour indexing over item iteration?

Reply:
I don't know :P. The body likely required access to the index at one point, but there's no reason to not use item iteration now.